cranelift/codegen/src/remove_constant_phis.rs

```diff
@@ -0,0 +1,387 @@
+//! A Constant-Phi-Node removal pass.
+
+use log::info;
+
+use crate::dominator_tree::DominatorTree;
+use crate::entity::EntityList;
+use crate::fx::FxHashMap;
+use crate::fx::FxHashSet;
+use crate::ir::instructions::BranchInfo;
+use crate::ir::Function;
+use crate::ir::{Block, Inst, Value};
+use crate::timing;
+
+use smallvec::{smallvec, SmallVec};
+use std::vec::Vec;
+
+// A note on notation.  For the sake of clarity, this file uses the phrase
+// "formal parameters" to mean the `Value`s listed in the block head, and
+// "actual parameters" to mean the `Value`s passed in a branch or a jump:
+//
+// block4(v16: i32, v18: i32):    <-- formal parameters
+//   ...
+//   brnz v27, block7(v22, v24)   <-- actual parameters
+//   jump block6
+
+// This transformation pass (conceptually) partitions all values in the
+// function into two groups:
+//
+// * Group A: values defined by block formal parameters, except for the entry block.
+//
+// * Group B: All other values: that is, values defined by instructions,
+//   and the formals of the entry block.
+//
+// For each value in Group A, it attempts to establish whether it will have
+// the value of exactly one member of Group B.  If so, the formal parameter is
+// deleted, all corresponding actual parameters (in jumps/branches to the
+// defining block) are deleted, and a rename is inserted.
+//
+// The entry block is special-cased because (1) we don't know what values flow
+// to its formals and (2) in any case we can't change its formals.
+//
+// Work proceeds in three phases.
+//
+// * Phase 1: examine all instructions.  For each block, make up a useful
+//   grab-bag of information, `BlockSummary`, that summarises the block's
+//   formals and jump/branch instruction.  This is used by Phases 2 and 3.
+//
+// * Phase 2: for each value in Group A, try to find a single Group B value
+//   that flows to it.  This is done using a classical iterative forward
+//   dataflow analysis over a simple constant-propagation style lattice.  It
+//   converges quickly in practice -- I have seen at most 4 iterations.  This
+//   is relatively cheap because the iteration is done over the
+//   `BlockSummary`s, and does not visit each instruction.  The resulting
+//   fixed point is stored in a `SolverState`.
+//
+// * Phase 3: using the `SolverState` and `BlockSummary`, edit the function to
+//   remove redundant formals and actuals, and to insert suitable renames.
+//
+// Note that the effectiveness of the analysis depends on on the fact that
+// there are no copy instructions in Cranelift's IR.  If there were, the
+// computation of `actual_absval` in Phase 2 would have to be extended to
+// chase through such copies.
+//
+// For large functions, the analysis cost using the new arm64 backend is about
```

Comment:
tiny, tiny nit: s/arm64/AArch64/ (we did a bulk find/replace during merge and have tried to standardize on the official name as far as we can)